### PR TITLE
[improve][broker] Should notify bundle ownership listener onLoad event when ServiceUnitState start (ExtensibleLoadManagerImpl only)

### PR DESCRIPTION
**Pull request by Demogorgon314**
### Motivation

The `ExtensibleLoadManagerImpl` behavior is not the same as the old LB, the old LB’s ownership will be empty when the broker starts.

When `ExtensibleLoadManagerImpl` starts, it may have some owners will be loaded into table view, and those ownerships will not trigger the bundle ownership listener, which may cause some issues, for example, the kop relies on the load event to load the offset topics.

### Modifications

Notify bundle ownership listener `onLoad` event when ServiceUnitState starts.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:335]: Actually, I think we can skip `stateChangeListeners.notify` as this is designed for active state change events.

can we run this after `tableview.listen` below? I think we better register `tableview.listen` asap.

**Review by heesung-sn — DISMISSED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: It feels like this should be using `tableview.forEachAndListen((key, value) -> handle(key, value));` instead?

Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: @heesung-sn @Demogorgon314 Please check this comment before merging.

Comment (heesung-sn) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java]: I dont think we should use foreachAndListen because we dont want the brokers to react on old messages. This can cause the unnecessary events in the channel.

Comment (Demogorgon314) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: Yes, we should use the current way, and we may need to impl an internal table view for our case, there is a race condition between the table view create and listen, so it may miss some events.

Comment (heesung-sn) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: ` there is a race condition between the table view create and listen, so it may miss some events`

I agree. Ideally, the table view constructor should accept the listeners to register them before receiving messages. I think forEachAndListen interface has the same problem too.

**Review by heesung-sn — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java
Comment (Demogorgon314) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: > I agree. Ideally, the table view constructor should accept the listeners to register them before receiving messages. I think forEachAndListen interface has the same problem too.

+1

Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: > I agree. Ideally, the table view constructor should accept the listeners to register them before receiving messages. I think forEachAndListen interface has the same problem too.

How do the changes in this PR fix the issue?
It seems odd to have some code duplication in the `.forEach` handler and only handle a specific case instead of all possible cases. 

Why don't we fix the root cause in the table view?

Comment (heesung-sn) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: > @heesung-sn How do the changes in this PR fix the issue?

No. this PR does not fix the racing condition issue. I think we can track this as a separate issue, here https://github.com/apache/pulsar/issues/23158


> It seems odd to have some code duplication in the .forEach handler and only handle a specific case instead of all possible cases.

This is because the current table view's message handlers cannot differentiate "existing" and "tail" messages. The proposed solution in`.foreach` in this PR is partially duplicate logic to separately handle "existing" messages. I also mentioned this issue here, https://github.com/apache/pulsar/issues/23158

> Why don't we fix the root cause in the table view?

I assume this may require a separate PIP to improve the TableView interface. 


Currently, in the worst case, the race condition(`between the table view create and listen, so it may miss some events`) can be resolved by the leader's bundle state channel monitor, which should fix any invalid state eventually.

Comment (heesung-sn) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: I am not sure if this is a blocker for this PR, but I raised a PIP for this TableView update, https://github.com/apache/pulsar/pull/23167

Comment (Demogorgon314) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: Should we merge this PR first and then after the TableView update is finished push another PR to fix the race condition issue?

Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: based on reading the current code of TableView, I don't yet understand why this PR isn't simply using `tableview.forEachAndListen((key, value) -> handle(key, value));`. I'll experiment to see if the tests would pass with that change.

Comment (heesung-sn) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: > I dont think we should use foreachAndListen because we dont want the brokers to react on old messages. This can cause the unnecessary events in the channel.

Let me give an example for this.

Currently, the topic has the following messages.
M1: bundle1, (Releasing from B1 to B2)
M2: bundle1, (Assigning to B2)
M3: bundle1, (Owned by B2)


When the topic does not run compaction yet, the table view can receive all non-compacted existing messages, M1-M3, when starting the table view on B1. Then, the broker B1 will unnecessarily close the Bundle, bundle1 if we pass 'handle()' to 'foreachAndListen'.

Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: @Demogorgon314 @heesung-sn In that case, I'd recommend refactoring the code so that `handle` is renamed to `handleEvent` and the function passed to the `.forEach` is moved to a new method `handleExisting` (or whatever naming describes this better).
Once the PIP for TableView improvements is processed, it can be migrated to use that. Makes sense?

Comment (heesung-sn) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java:337]: I agree with this change.

- handle() -> handleEvent()
- the lambda passed to foreach -> handleExisting()

And

> Once the PIP for TableView improvements is processed, it can be migrated to use that. Makes sense?

I agree.

Comment (Demogorgon314) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java]: Updated.

**Review by lhotari — APPROVED** [review]
LGTM

**Comment by lhotari** [comment]
This test is very flaky. Would it be possible to fix it or disable it until it works? Reported #23240 for this

**Pull request merged**
